src/promnesia/sources/reddit.py

```diff
@@ -3,40 +3,153 @@
 '''
 
 from itertools import chain
+from typing import Set
 
 from ..common import Visit, Loc, extract_urls, Results, logger
 
 
-def index() -> Results:
+# mostly here so we can keep track of how the user
+# wants to render markdown
+class RedditRenderer:
+
+    def __init__(self, render_markdown: bool = False):
+
+        try:
+            from .markdown import TextParser, extract_from_text
+            self._link_extractor = extract_from_text
+            self._parser_cls = TextParser
+        except ImportError as import_err:
+            # TODO: add dummy _link_extractor and _parser_cls classes incase
+            # these are called by a subclass?
+
+            # only send error if the user is trying to enable this feature
+            if render_markdown:
+                logger.exception(import_err)
+                logger.critical("Could not import markdown module to render reddit markdown. Try 'python3 -m pip install mistletoe'")
+            render_markdown = False  # force to be false, couldn't import
+        self.render_markdown = render_markdown
+
+
+    def _from_comment(self, i: 'Comment') -> Results:
+        locator = Loc.make(
+            title='Reddit comment',
+            href=i.url,
+        )
+        yield from self._from_common(i, locator=locator)
+
+
+    def _from_submission(self, i: 'Submission') -> Results:
+        locator = Loc.make(
+            title=f'Reddit submission: {i.title}',
+            href=i.url,
+        )
+        yield from self._from_common(i, locator=locator)
+
+
+    def _from_upvote(self, i: 'Upvote') -> Results:
+        locator = Loc.make(
+            title=f'Reddit upvote',
+            href=i.url,
+        )
+        yield from self._from_common(i, locator=locator)
+
+
+    def _from_save(self, i: 'Save') -> Results:
+        locator = Loc.make(
+            title='Reddit save',
+            href=i.url,
+        )
+        yield from self._from_common(i, locator=locator)
+
+
+    # to allow for possible subclassing by the user?
+    def _render_body(self, text: str) -> str:
+        if self.render_markdown:
+            return self._parser_cls(text)._doc_ashtml()
```

Comment:
imo a bit nasty when class has different attributes depending on constructor parameters, harder for both people and fo mypy
maybe instead make `_parser_cls` optional and check if it's `None`?